controllers/lease_controller.go

```diff
@@ -0,0 +1,456 @@
+	batchv1 "k8s.io/api/batch/v1"
+	coordinationv1 "k8s.io/api/coordination/v1"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	errorsutil "k8s.io/apimachinery/pkg/util/errors"
+	"k8s.io/utils/pointer"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+
+	druidv1alpha1 "github.com/gardener/etcd-druid/api/v1alpha1"
+	controllersconfig "github.com/gardener/etcd-druid/controllers/config"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+)
+
+// LeaseController reconciles compaction job
+type LeaseController struct {
+	client.Client
+	Scheme *runtime.Scheme
+	logger logr.Logger
+	config controllersconfig.CompactionConfig
+}
+
+// NewLeaseController creates a new LeaseController object
+func NewLeaseController(mgr manager.Manager, config controllersconfig.CompactionConfig) *LeaseController {
+	return &LeaseController{
+		Client: mgr.GetClient(),
+		Scheme: mgr.GetScheme(),
+		logger: log.Log.WithName("lease-controller"),
+		config: config,
+	}
+}
+
+// +kubebuilder:rbac:groups=druid.gardener.cloud,resources=etcds,verbs=get;list;watch
+// +kubebuilder:rbac:groups=coordination.k8s.io,resources=leases,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=batch,resources=jobs,verbs=get;create;list;watch;update;patch;delete
+
+// Reconcile reconciles the compaction job.
+func (lc *LeaseController) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	lc.logger.Info("Lease controller reconciliation started")
+	etcd := &druidv1alpha1.Etcd{}
+	if err := lc.Get(ctx, req.NamespacedName, etcd); err != nil {
+		if errors.IsNotFound(err) {
+			// Object not found, return. Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return ctrl.Result{
+				RequeueAfter: 10 * time.Second,
+			}, err
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{
+			RequeueAfter: 10 * time.Second,
+		}, err
+	}
+
+	logger := lc.logger.WithValues("etcd", kutil.Key(etcd.Namespace, etcd.Name).String())
+
+	// Get or Create delta_snapshot_revisions lease object that will keep track of delta snapshot revisions based on which
+	// compaction job will be scheduled
+	nsName := types.NamespacedName{
+		Name:      getDeltaLease(etcd),
+		Namespace: etcd.Namespace,
+	}
+
+	deltaLease := &coordinationv1.Lease{}
+	err := lc.Get(ctx, nsName, deltaLease)
+	if err != nil {
+		logger.Info("Couldn't fetch delta snap lease because: " + err.Error())
+
+		return ctrl.Result{
+			RequeueAfter: 10 * time.Second,
+		}, err
+	}
+
+	// Run compaction job
+	if etcd.Spec.Backup.Store != nil {
+		hi, err := strconv.ParseInt(*deltaLease.Spec.HolderIdentity, 10, 32)
```

Comment:
Would suggest using a 64 bit int here instead of a 32 bit int as `Snapshot.LastRevision`s are `int64` type